src/OpenTK.Windowing.Common/Input/IJoystickState.cs

```diff
@@ -0,0 +1,81 @@
+﻿//
+// IJoystickState.cs
+//
+// Copyright (C) OpenTK
+//
+// This software may be modified and distributed under the terms
+// of the MIT license. See the LICENSE file for details.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace OpenTK.Windowing.Common.Input
+{
+    /// <summary>
+    /// Encapsulates the state of a joystick device.
+    /// </summary>{
+    public interface IJoystickState
```

Comment:
Why do we need an interface for this?

Reply:
Good question: provided all mutation is done by internal or private members, the interface is strictly speaking unnecessary. However, the move of all input closer to GraphicsLibraryFramework (where GamePadState already exists) represents a shift that we might want to keep consistent: Should we be facading GLFW away from NativeWindow more generally? Do we foresee Something Else in the place of GLFW? Would it be useful to have the option when we get to Vulkan?

Reply:
The original reason behind the interface was that it did not include the `Snapshot()` method. This separated state from behaviour as you saw above.

Reply:
@varon So then my question would be how the end dev/user is expected to Snapshot the JoystickState: the JoystickState class is internal, and so the `Snapshot()` method would be hidden. Even if JoystickState were public, the method wouldn't be obviously available without a cast.

Reply:
Most of the issues surrounding encapsulation stem from the `INativeWindow` interface existing. We only have one implementation of it and as @NogginBops asks, "Why do we need an interface for this?".

Definitely it was a victim of premature generalization, which is cascading into this issue.

If we dropped the (as far as I can see pointless) `INativeWindow` interface we could expose `JoystickState` directly which can include the `Snapshot()` method.

Reply:
Great, let's take the opportunity now to do that.